text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,364 @@
+
+```rust
+let mut v: ArrayVec<i32, 4> = ArrayVec::new();
+v.push(1);
+v.push(2);
+
+assert_eq!(v.len(), 2);
+assert_eq!(v[0], 1);
+
+assert_eq!(v.pop(), Some(2));
+assert_eq!(v.len(), 1);
+
+v[0] = 7;
+assert_eq!(v[0], 7);
+
+v.extend([1, 2, 3].iter().copied());
+
+for element in &v {
+    println!("{}", element);
+}
+assert_eq!(v, [7, 1, 2, 3]);
+```
+
+Instead of relying on a heap-allocator, stack-based memory is added and removed on-demand in a last-in-first-out (LIFO) order according to the calling workflow of a program. `ArrayVec` takes advantage of this predictable behavior to reserve an exactly amount of uninitialized bytes up-front to form an internal buffer.
+
+```rust
+// `array_vec` can store up to 64 elements
+let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+```
+
+Another potential use-case is the usage within constant environments:
+
+```rust
+const MY_CONST_ARRAY_VEC: ArrayVec<i32, 10> = {
+    let mut v = ArrayVec::new();
+    v.push(1);
+    v.push(2);
+    v.push(3);
+    v.push(4);
+    v
+};
+```
+
+Of course, fixed buffers lead to inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+```rust
+// This vector can store up to 0 elements, therefore, nothing at all
+let mut array_vec: ArrayVec<i32, 0> = ArrayVec::new();
+array_vec.push(1); // Error!
+```
+
+A good question is: Should I use `core::collections::ArrayVec<T>` or `alloc::vec::Vec<T>`? Well, `Vec` is already good enough for most situations while stack allocation usually shines for small sizes.
+
+* Do you have a known upper bound?
+
+* How much memory are you going to allocate for your program? The default values of `RUST_MIN_STACK` or `ulimit -s` might not be enough.
+
+* Are you using nested `Vec`s? `Vec<ArrayVec<T, N>>` might be better than `Vec<Vec<T>>`.
+
+```
+let _: Vec<Vec<i32>> = vec![vec![1, 2, 3], vec![4, 5]];
+
+ +-----+-----+-----+
+ | ptr | len | cap |
+ +--|--+-----+-----+
+    |
+    |   +---------------------+---------------------+----------+
+    |   |       Vec<i32>      |       Vec<i32>      |          |
+    |   | +-----+-----+-----+ | +-----+-----+-----+ |  Unused  |
+    '-> | | ptr | len | cap | | | ptr | len | cap | | capacity |
+        | +--|--+-----+-----+ | +--|--+-----+-----+ |          |
+        +----|----------------+----|----------------+----------+
+             |                     |
+             |                     |   +---+---+--------+
+             |                     '-> | 4 | 5 | Unused |
+             |                         +---+---+--------+
+             |   +---+---+---+--------+
+             '-> | 1 | 2 | 3 | Unused |
+                 +---+---+---+--------+
+
+Illustration credits: @mbartlett21
+```
+
+Can you see the `N`, where `N` is length of the external `Vec`, calls to the heap allocator? In the following illustration, the internal `ArrayVec`s are placed contiguously in the same space.
```

Comment:
This could possibly be worded better:

As shown in the diagram above, each element of the vector has its own separate heap allocation, meaning more book-keeping for the global allocator to do.